src/analysis/symbols.rs

```diff
@@ -0,0 +1,147 @@
+use analysis::namespaces::{self, NsId};
+use case::CaseExt;
+use library::*;
+use std::collections::HashMap;
+
+#[derive(Clone, Debug, Default)]
+pub struct Symbol {
+    crate_name: Option<String>,
+    owner_name: Option<String>,
+    name: String,
+}
+
+impl Symbol {
+    pub fn full_rust_name(&self) -> String {
+        let mut ret = String::new();
+        if let Some(ref s) = self.crate_name {
+            ret.push_str(s);
+            ret.push_str("::");
+        }
+        if let Some(ref s) = self.owner_name {
+            ret.push_str(s);
+            ret.push_str("::");
+        }
+        ret.push_str(&self.name);
+        ret
+    }
+
+    pub fn make_trait_method(&mut self) {
+        let name = self.owner_name.take();
+        self.owner_name = name.map(|s| format!("{}Ext", s));
+    }
+}
+
+#[derive(Debug)]
+pub struct Info {
+    symbols: Vec<Symbol>,
+    c_name_index: HashMap<String, u32>,
+}
+
+pub fn run(library: &Library, namespaces: &namespaces::Info) -> Info {
+    let mut info = Info {
+        symbols: Vec::new(),
+        c_name_index: HashMap::new(),
+    };
+
+    info.insert("NULL", Symbol { name: "None".into(), ..Default::default() });
+    info.insert("FALSE", Symbol { name: "false".into(), ..Default::default() });
+    info.insert("TRUE", Symbol { name: "true".into(), ..Default::default() });
+
+    for (ns_id, ref ns) in library.namespaces.iter().enumerate() {
+        let ns_id = ns_id as NsId;
+        if ns_id == namespaces::INTERNAL {
+            continue
+        }
+
+        let crate_name;
+        if ns_id == namespaces::MAIN {
+            crate_name = None;
+        }
+        else {
+            crate_name = Some(&namespaces[ns_id].crate_name);
+        }
```

Comment:
Okay looks like this style

```rust
        let crate_name = if ns_id == namespaces::MAIN {
            None
        } else {
            Some(&namespaces[ns_id].crate_name)
        };
```
is all over rustc and libstd code so I should get used to it.